File: internal/shadow/adapters/git.go
package adapters

import (
	"fmt"
	"os"
	"strings"
)

func gitCapture(args []string, dir string) (string, error) {
	return Capture("git", args, ExecOptions{Dir: dir})
}

// CloneShallow shallow-clones a repo at a ref into dir, authenticated with a token.
func CloneShallow(repo, ref, dir, token string) error {
	url := fmt.Sprintf("https://x-access-token:%s@github.com/%s.git", token, repo)
	return Run("git", []string{"clone", "--depth=1", "--branch", ref, url, dir}, ExecOptions{})
}

// ConfigureBotIdentity sets a committer identity for the shadow bot commit.
func ConfigureBotIdentity(dir string) error {
	if _, err := gitCapture([]string{"config", "user.name", "shadow-testing[bot]"}, dir); err != nil {
		return err
	}
	_, err := gitCapture([]string{"config", "user.email", "shadow-testing@users.noreply.github.com"}, dir)
	return err
}

// ResetBranchToEmptyTree starts the shadow branch from HEAD and clears the tracked tree, so the
// next commit contains only the mirrored consumer files.
func ResetBranchToEmptyTree(dir, branch string) error {
	if _, err := gitCapture([]string{"checkout", "-B", branch}, dir); err != nil {
		return err
	}
	_, err := gitCapture([]string{"rm", "-rf", "--quiet", "."}, dir)
	return err
}

// deterministicDate fixes the commit timestamp so identical (tree, parent, message, identity)
// yields an identical SHA — re-runs are no-ops (force-push is idempotent, the prior run is reused).
const deterministicDate = "2000-01-01T00:00:00Z"

// CommitAll stages everything and makes a reproducible commit.
func CommitAll(dir, message string) error {
	if _, err := gitCapture([]string{"add", "-A"}, dir); err != nil {
		return err
	}
	env := append(os.Environ(), "GIT_AUTHOR_DATE="+deterministicDate, "GIT_COMMITTER_DATE="+deterministicDate)
	_, err := Capture("git", []string{"commit", "--allow-empty", "-m", message}, ExecOptions{Dir: dir, Env: env})
	return err
}

// ForcePush force-pushes the branch to origin.
func ForcePush(dir, branch string) error {
	_, err := gitCapture([]string{"push", "--force", "origin", branch}, dir)
	return err
}

// HeadSha returns the current HEAD commit SHA.
func HeadSha(dir string) (string, error) {
	out, err := gitCapture([]string{"rev-parse", "HEAD"}, dir)
	return strings.TrimSpace(out), err
}

File: shadow/src/adapters/git.ts
import { capture, run } from './exec.ts';

const git = (args: string[], cwd?: string) => capture('git', args, { cwd });

/** Shallow-clone a consumer repo at a ref into `dir`, authenticated with a token. */
export async function cloneShallow(opts: {
  repo: string;
  ref: string;
  dir: string;
  token: string;
}): Promise<void> {
  const url = `https://x-access-token:${opts.token}@github.com/${opts.repo}.git`;
  await run('git', ['clone', '--depth=1', '--branch', opts.ref, url, opts.dir]);
}

/** Set a committer identity for the bot commit. */
export async function configureBotIdentity(cwd: string): Promise<void> {
  await git(['config', 'user.name', 'shadow-testing[bot]'], cwd);
  await git(['config', 'user.email', 'shadow-testing@users.noreply.github.com'], cwd);
}

/** Start the shadow branch from the current HEAD and clear the tracked tree, so the next commit
 * contains only the mirrored consumer files (a clean single commit on top of the runner base). */
export async function resetBranchToEmptyTree(cwd: string, branch: string): Promise<void> {
  await git(['checkout', '-B', branch], cwd);
  await git(['rm', '-rf', '--quiet', '.'], cwd);
}

// Fixed timestamp so an identical (tree, parent, message, identity) yields an identical commit SHA.
// Combined with the fixed bot identity, this makes the shadow commit reproducible: re-running with
// the same inputs produces the same SHA, the force-push is a no-op, and we observe the already-
// concluded run instead of spawning a redundant one.
const DETERMINISTIC_DATE = '2000-01-01T00:00:00Z';

/** Stage everything and make a reproducible commit. */
export async function commitAll(cwd: string, message: string): Promise<void> {
  await git(['add', '-A'], cwd);
  await capture('git', ['commit', '--allow-empty', '-m', message], {
    cwd,
    env: { ...process.env, GIT_AUTHOR_DATE: DETERMINISTIC_DATE, GIT_COMMITTER_DATE: DETERMINISTIC_DATE },
  });
}

export async function forcePush(cwd: string, branch: string): Promise<void> {
  await git(['push', '--force', 'origin', branch], cwd);
}

/** The current HEAD commit SHA — used to watch the exact run for this push. */
export async function headSha(cwd: string): Promise<string> {
  return (await git(['rev-parse', 'HEAD'], cwd)).trim();
}
